File: libs/api/src/main/java/com/example/api/Tls12SslSocketFactory.java
package com.example.api;

import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;

import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;

import okhttp3.TlsVersion;

public class Tls12SslSocketFactory extends SSLSocketFactory {
	private final String PROTOCOL = TlsVersion.TLS_1_2.toString();
	private SSLSocketFactory socketFactory;

	Tls12SslSocketFactory() {
		socketFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
	}

	@Override
	public String[] getDefaultCipherSuites() {
		return socketFactory.getDefaultCipherSuites();
	}

	@Override
	public String[] getSupportedCipherSuites() {
		return socketFactory.getSupportedCipherSuites();
	}

	@Override
	public Socket createSocket(Socket socket, String host, int port, boolean autoClose) throws IOException {
		SSLSocket sslSocket = (SSLSocket) socketFactory.createSocket(socket, host, port, autoClose);
		sslSocket.setEnabledProtocols(new String[]{PROTOCOL});

		return sslSocket;
	}

	@Override
	public Socket createSocket(String host, int port) throws IOException {
		SSLSocket sslSocket = (SSLSocket) socketFactory.createSocket(host, port);
		sslSocket.setEnabledProtocols(new String[]{PROTOCOL});

		return sslSocket;
	}

	@Override
	public Socket createSocket(String host, int port, InetAddress localHost, int localPort) throws IOException {
		SSLSocket sslSocket = (SSLSocket) socketFactory.createSocket(host, port, localHost, localPort);
		sslSocket.setEnabledProtocols(new String[]{PROTOCOL});

		return sslSocket;
	}

	@Override
	public Socket createSocket(InetAddress host, int port) throws IOException {
		SSLSocket sslSocket = (SSLSocket) socketFactory.createSocket(host, port);
		sslSocket.setEnabledProtocols(new String[]{PROTOCOL});

		return sslSocket;
	}

	@Override
	public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort) throws IOException {
		SSLSocket sslSocket = (SSLSocket) socketFactory.createSocket(address, port, localAddress, localPort);
		sslSocket.setEnabledProtocols(new String[]{PROTOCOL});

		return sslSocket;
	}
}

File: libs/api/src/main/kotlin/com/example/api/Tls12SslSocketFactory.kt
package com.example.api

import okhttp3.TlsVersion
import java.io.IOException
import java.net.InetAddress
import java.net.Socket
import javax.net.ssl.SSLSocket
import javax.net.ssl.SSLSocketFactory

class Tls12SslSocketFactory internal constructor() : SSLSocketFactory() {
  companion object {
    private val PROTOCOL = TlsVersion.TLS_1_2.toString()
  }

  private val socketFactory: SSLSocketFactory = SSLSocketFactory.getDefault() as SSLSocketFactory

  override fun getDefaultCipherSuites(): Array<String> {
    return socketFactory.defaultCipherSuites
  }

  override fun getSupportedCipherSuites(): Array<String> {
    return socketFactory.supportedCipherSuites
  }

  @Throws(IOException::class)
  override fun createSocket(socket: Socket, host: String, port: Int, autoClose: Boolean): Socket {
    val sslSocket = socketFactory.createSocket(socket, host, port, autoClose) as SSLSocket
    sslSocket.enabledProtocols = arrayOf(PROTOCOL)

    return sslSocket
  }

  @Throws(IOException::class)
  override fun createSocket(host: String, port: Int): Socket {
    val sslSocket = socketFactory.createSocket(host, port) as SSLSocket
    sslSocket.enabledProtocols = arrayOf(PROTOCOL)

    return sslSocket
  }

  @Throws(IOException::class)
  override fun createSocket(host: String, port: Int, localHost: InetAddress, localPort: Int): Socket {
    val sslSocket = socketFactory.createSocket(host, port, localHost, localPort) as SSLSocket
    sslSocket.enabledProtocols = arrayOf(PROTOCOL)

    return sslSocket
  }

  @Throws(IOException::class)
  override fun createSocket(host: InetAddress, port: Int): Socket {
    val sslSocket = socketFactory.createSocket(host, port) as SSLSocket
    sslSocket.enabledProtocols = arrayOf(PROTOCOL)

    return sslSocket
  }

  @Throws(IOException::class)
  override fun createSocket(
    address: InetAddress,
    port: Int,
    localAddress: InetAddress,
    localPort: Int
  ): Socket {
    val sslSocket = socketFactory.createSocket(address, port, localAddress, localPort) as SSLSocket
    sslSocket.enabledProtocols = arrayOf(PROTOCOL)

    return sslSocket
  }
}
